### PR TITLE
Favorites API.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi Max,

I wanted to clean it up a bit; will do it a day or two later though. This is what is now in production.

PTAL.

Thanks!
Dima

### CODE REVIEW

File: schema.h [hunk 1]
```diff
@@ -194,6 +194,10 @@ struct User : yoda::Padawan {
   template <typename A>
   void serialize(A& ar) {
     Padawan::serialize(ar);
+    // TODO(dkorolev): We don't have UID here.
+    // If it's OK, remove CID from another place too.
+    // If it's not OK, fix it.
+    // Not urgent since we can -- and will soon! -- replay the log to update the DB!
```
Comment (mzhurovich) [inline schema.h:200]: Looks like it's not ok :-/

File: schema.h [hunk 2]
```diff
@@ -308,6 +312,26 @@ struct Answer : yoda::Padawan {
...
 
+struct Favorite : yoda::Padawan {
+  UID uid = UID::INVALID;
+  CID cid = CID::INVALID;
+  bool favorited = false;
+
+  Favorite() = default;
+  Favorite(const UID uid, const CID cid, bool favorited = false) : uid(uid), cid(cid), favorited(favorited) {}
+
+  const UID& row() const { return uid; }
```
Comment (mzhurovich) [inline schema.h]: Why not return by value as `CID col()` below?

Comment (dkorolev) [inline schema.h]: :+1: f6f95b4

File: schema.h (hunk 1 above)
Comment (dkorolev) [inline schema.h:200]: Not a problem -- a log replay tool is already in the drafts.

**Comment by dkorolev** [comment]
@mzhurovich Thanks! Fix is in :-)

File: server.h [hunk 3]
```diff
@@ -206,6 +206,88 @@ class CTFOServer final {
...
+                    }
+                  } catch (yoda::SubscriptException<Favorite>) {
+                    // No favorites for this user.
+                  }
+
+                  // In reverse chronological order.
+                  std::sort(favs.rbegin(), favs.rend());
+
+                  // And publish them.
+                  const auto GenerateCardForFavs = [this, uid, &favorites](const Card& card) {
```
Comment (mzhurovich) [inline server.h]: Do you need `uid` and `&favorites` here?

File: server.h [hunk 4]
```diff
@@ -277,6 +364,15 @@ class CTFOServer final {
       card_entry.ctfo_count = card.ctfo_count;
       card_entry.tfu_count = card.tfu_count;
       card_entry.skip_count = card.skip_count;
+      card_entry.favorited = false;
+      try {
+        const EntryWrapper<Favorite> fav = favorites.Get(uid, card.cid);
```
Comment (mzhurovich) [inline server.h]: Isn't `try/catch` redundant here?

Comment (dkorolev) [inline server.h]: Yes! Removed!

You made me curious to look into, and after 1.5 hours of looking through stack traces like  http://pastebin.com/dTrF20LE, and adding https://github.com/c5t/Current/pull/231, it turned out to be a missing Cereal's registerer. Thanks!

File: server.h (hunk 3 above)
Comment (dkorolev) [inline server.h]: Nope, removed in 41c584d. Thanks! :+1:

**Comment by dkorolev** [comment]
@mzhurovich : Fixed, thanks again for great notes. PTAL.

Notes to self: https://gist.github.com/dkorolev/30546e75e608f3ee410b

Thanks!
Dima

**Comment by dkorolev** [comment]
Thank you Max :+1:

**Pull request merged**